### ISSUE TITLE
Warning: Message parser reports malformed message packet.

### ISSUE THREAD

**Issue by ptorrent** [song940/node-dns#8]
Hello,

When i execute : "dig wbc.weble.ch +trace"

I've this result : 

```
; <<>> DiG 9.10.3-P4-Raspbian <<>> wbc.weble.ch +trace
;; global options: +cmd
.                       72776   IN      NS      e.root-servers.net.
.                       72776   IN      NS      f.root-servers.net.
.                       72776   IN      NS      g.root-servers.net.
.                       72776   IN      NS      h.root-servers.net.
.                       72776   IN      NS      i.root-servers.net.
.                       72776   IN      NS      j.root-servers.net.
.                       72776   IN      NS      k.root-servers.net.
.                       72776   IN      NS      l.root-servers.net.
.                       72776   IN      NS      m.root-servers.net.
.                       72776   IN      NS      a.root-servers.net.
.                       72776   IN      NS      b.root-servers.net.
.                       72776   IN      NS      c.root-servers.net.
.                       72776   IN      NS      d.root-servers.net.
;; Received 811 bytes from 172.16.1.1#53(172.16.1.1) in 3 ms

ch.                     172800  IN      NS      a.nic.ch.
ch.                     172800  IN      NS      b.nic.ch.
ch.                     172800  IN      NS      c.nic.ch.
ch.                     172800  IN      NS      d.nic.ch.
ch.                     172800  IN      NS      e.nic.ch.
ch.                     172800  IN      NS      f.nic.ch.
ch.                     172800  IN      NS      g.nic.ch.
ch.                     172800  IN      NS      h.nic.ch.
ch.                     86400   IN      DS      11896 13 2 24EE6537B1C452D3AEBF439DCF74024717054152DA7F206D5FCBA1A9 0F70711F
ch.                     86400   IN      RRSIG   DS 8 1 86400 20190930050000 20190917040000 59944 . kaGRIHYMMuruHaWAIHb9kVXZNH9tzDNkqgQcX9oFotbd86puoY6/sRp6 +y8lhBxUgm7qi7GM3JxTZ3iZ3/dpgNv7PTVLskDUwrscSQC5jkxzOzvB 5oo59keBmC9X2UlJ+rLJoLNLNuao+6LyMPZtSgW7DudvN1rMQpZM39is XuG5URCG9qSBzkJt3FPESMBT+h356sbvN4Jz8GvwPLaE2mpdNnuBwAIG qBYaqDiWhE0pLuSVwdLlPftCnJ7RT1dNfTI3wJfSl2fDI6TpHPgr8bSt 4H5HDutqvsZuTQp1p2CqKLtAJOCIKBPOfact7SyYaVFlmuqhQMZLwFDP UnZDiQ==
;; Received 860 bytes from 193.0.14.129#53(k.root-servers.net) in 11 ms

weble.ch.               3600    IN      NS      ns1.switchplus.ch.
weble.ch.               3600    IN      NS      ns2.switchplus.ch.
HS53IQDLBM00MIRD1LPS0J47BPJP5SF8.ch. 900 IN NSEC3 1 1 2 87E195EB HS63KNDB8F15206UBJQDIRIISQ1EK7N7 NS SOA RRSIG DNSKEY NSEC3PARAM
HS53IQDLBM00MIRD1LPS0J47BPJP5SF8.ch. 900 IN RRSIG NSEC3 13 2 900 20191013161405 20190913160114 7967 ch. GZsRnHBemsdnY1WI6R9+O9NjL9a7s479t3Oqcw6j28LVsYZWOjqUPnSK bjRHaSKp4hZrtxX9gMzIRAJChgL3AQ==
LBHUG80GC3VGPEU435F60L91DBJ3LJE8.ch. 900 IN NSEC3 1 1 2 87E195EB LBK2U02LTLKUCOCKTKJG1O2EMMOK0HBB NS DS RRSIG
LBHUG80GC3VGPEU435F60L91DBJ3LJE8.ch. 900 IN RRSIG NSEC3 13 2 900 20191010205442 20190911113001 7967 ch. UdhiNXu1UZx2JE1qg/WI+I51OMtCanyanuqtArGcuMc+GyFZR7NeLK9l GpMKl5wmdz3EDm+TL+y8X7XWncNf8w==
;; Received 487 bytes from 194.146.106.10#53(f.nic.ch) in 13 ms

wbc.weble.ch.           3600    IN      NS      ns1.weble.ch.
wbc.weble.ch.           3600    IN      NS      ns2.weble.ch.
;; Received 109 bytes from 193.223.77.3#53(ns2.switchplus.ch) in 119 ms

**;; Warning: Message parser reports malformed message packet.**
wbc.weble.ch.           300     IN      A       213.193.121.210
;; Received 67 bytes from 213.193.121.211#53(ns1.weble.ch) in 31 ms
```

Here the response object : 
```javascript

var response = new dns.Packet(request);
	response.header.qr = 1;


	response.answers.push({
		name: oDomain,
		address: target.target,
		type	: dns.Packet.TYPE.A,
		ttl 	: ttl,
		class	: dns.Packet.CLASS.IN
	});
```

And in console I've:
![image](https://user-images.githubusercontent.com/6083644/65044508-4e476b00-d95d-11e9-903b-61205df7cde5.png)

Is there a link between this error and the warning **;; Warning: Message parser reports malformed message packet.** ?

Is there something wrong ?

**Comment by andsens** [comment]
I'm seeing the exact same issue. I'm a bit down the rabbit hole from minikube ingress-dns to `cryptexlabs/minikube-ingress-dns:0.2.1`, [to the sourcecode](https://gitlab.com/cryptexlabs/public/development/minikube-ingress-dns/-/blob/master/app/dns-server/nodejs/src/index.js), to this package.

My issue arises when the systemd-resolverd stub-resolver performs a query. Using `dig` to query the server directly is no problem. Here's the wireshark disector output:
![Screenshot from 2020-03-16 12-48-57](https://user-images.githubusercontent.com/342823/76755662-87359080-6784-11ea-8424-213a478c8632.png)

**Comment by andris9** [comment]
This happens when dns2 tries to answer to a request it did not understand (mostly OPT RR). The solution in my case was to filter out unsupported types from prepared responses.

```js
// create a filter of supported data types
const SUPPORTED_TYPES = new Set(
  Object.keys(dns.Packet.TYPE)
    .map((key) => dns.Packet.TYPE[key])
    .filter((val) => typeof val === "number")
);

const server = dns.createServer(function (request, send) {
	// filter out unsupported requests (eg. EDNS)
	if (request.additionals && request.additionals.length) {
	  request.additionals = request.additionals.filter((additional) =>
	    SUPPORTED_TYPES.has(additional.type)
	  );
	}
	....
```

**Comment by song940** [comment]
Simply clears out the `additionals` field in the response:

`response.additionals = [];`

**Comment by ardarico** [comment]
Thank you. 

response.additionals = []; removed the error message about unknown encoder, but the "unknown parser type: undefined(41) is still there.

**Comment by andsens** [comment]
> but the "unknown parser type: undefined(41) is still there.

That's because the encoder for that packet type doesn't exist in dns2.  
[Check my PR on gitlab for a more in-depth explanation](https://gitlab.com/cryptexlabs/public/development/minikube-ingress-dns/-/merge_requests/2).  
TL;DR: This mostly (exclusively?) happens when the DNS client sends edns queries (DNS with some crypto auth request attached).

**Comment by ardarico** [comment]
Thank you for the reply. I use dig as client and I would like to send the following request:

dig @127.0.0.1 -p 53 6gkzwgjz._iot._udp.unipr.it -t ptr

In my dns2-based server the response to a request with type PTR is created in the following way:

response.header.qr = 1;
response.header.ra = 1;
response.additionals = [];
response.answers.push({
          name: query,
          type: Packet.TYPE.PTR, 
          class: Packet.CLASS.IN,
          ttl: 100,
          address: names[query]
        });

where address is not actually an address but a name (6gkzwgjz.unipr.it in the considered example).

However, send(response)  fails with unknown encoder PTR(12). Now, PTR seems to be a supported type in dns2. I looked into the packet.js code but I don't have idea of how to fix this (I would like to).

**Comment by bambusoft** [comment]
Hi:
Running the example server for IPv6 I recieve => Warning: Message parser reports malformed message packet.

`
dwuser@darwin:~$ dig @localhost -p 5333 google.com aaaa
;; Warning: Message parser reports malformed message packet.

; <<>> DiG 9.16.1-Ubuntu <<>> @localhost -p 5333 google.com aaaa
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 1036
;; flags: qr rd ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available
;; WARNING: Message has 10 extra bytes at end

;; QUESTION SECTION:
;google.com.                    IN      AAAA

;; Query time: 8 msec
;; SERVER: 127.0.0.1#5334(127.0.0.1)
;; WHEN: Sun Mar 21 14:01:50 CST 2021
;; MSG SIZE  rcvd: 60
`
This is the coda I m using:
`
const dns2 = require('dns2');
const { Packet } = dns2;
const server = dns2.createUDPServer((request, send, rinfo) => {
  const response = Packet.createResponseFromRequest(request);
  const [ question ] = request.questions;
  const { name } = question;
  response.additionals = [];
  response.answers.push({
    name,
    type: Packet.TYPE.AAAA,
    class: Packet.CLASS.IN,
    ttl: 300,
    address: '2001:4860:4860::8888'
  });
  send(response);
});

server.on('request', (request, response, rinfo) => {
  console.log(request.header.id, request.questions[0]);
});
server.listen(5333);
`
Thank you
